### ISSUE TITLE
Wrong type with Promise chaining

### ISSUE THREAD

**Issue by tkrotoff** [Microsoft/TypeScript#12409]
With 2.1.4 (edit: first reported with typescript@2.2.0-dev.20161121):

```TypeScript
declare function createHello(): Promise<Hello>;

interface Hello {
  world(): Promise<any>;
}

createHello()
  .then(hello => hello.world())
  .then(message => {
    console.log(message); // BUG: message is of type Hello instead of any
  });
```

On TypeScript playground, no problem (edit: this is not true anymore, was working with 2.0.10, not with 2.1.4): https://www.typescriptlang.org/play/index.html#src=declare%20function%20createHello()%3A%20Promise%3CHello%3E%3B%0D%0A%0D%0Ainterface%20Hello%20%7B%0D%0A%20%20world()%3A%20Promise%3Cany%3E%3B%0D%0A%7D%0D%0A%0D%0AcreateHello()%0D%0A%20%20.then(hello%20%3D%3E%20hello.world())%0D%0A%20%20.then(message%20%3D%3E%20%7B%0D%0A%20%20%20%20console.log(message)%3B%20%2F%2F%20message%20is%20of%20type%20any%20%3D%3E%20works%0D%0A%20%20%7D)%3B%0D%0A

(for the little story, the bug was found here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/12808#issuecomment-261752232)

**Comment by RohithKumbharkar** [comment]
@tkrotoff I'm not sure if it's an _issue_ or a _feature_ but the recent changes to `then` callback might be **breaking code** & the alternative I see is to use the generic version of `then<TResult>`.

```typescript
createHello()
  .then<any>(hello => hello.world())
  .then(message => {
    console.log(message); // Now message is of type any
});
```

[Breaking changes](http://stackoverflow.com/a/40822110/5211120)

**Comment by jwbay** [comment]
Duplicate of #10977 I think

**Comment by tkrotoff** [comment]
TypeScript 2.2 Promise implementation is still broken, example:
```TypeScript
interface Toto {}

const promise = new Promise<Toto>((resolve, reject) => {
  resolve('toto');
});

promise
  .then(value => { // value is of type Toto => OK
    console.log('then1 value=', value);
    return 'titi';
  })
  .then(value => { // value is of type Toto => KO, should be 'titi' or string
    console.log('then2 value=', value);
    return 2;
  })
  .then(value => { // value is of type Toto => KO, should be 2 or number
    console.log('then3 value=', value);
  });
```

Verification by executing the code inside Chrome DevTools:
```
then1 value= toto
then2 value= titi
then3 value= 2
```

**Comment by jwbay** [comment]
The fix is in version 2.3, not 2.2.
